django_mongodb/compiler.py

```diff
@@ -0,0 +1,348 @@
+from functools import cached_property
+from itertools import chain
+
+from django.db.models.sql.compiler import (
+    SQLCompiler as BaseSQLCompiler,
+)
+from django.db.models.sql.compiler import (
+    SQLInsertCompiler as BaseSQLInsertCompiler,
+)
+from django.db.models.sql.constants import (
+    CURSOR,
+    GET_ITERATOR_CHUNK_SIZE,
+    MULTI,
+    NO_RESULTS,
+    SINGLE,
+)
+from pymongo import InsertOne, UpdateOne
+
+from django_mongodb.query import MongoOrdering, MongoSelect, MongoWhereNode
+
+
+class SQLCompiler(BaseSQLCompiler):
+    def __init__(self, query, connection, using, elide_empty=True):
+        super().__init__(query, connection, using, elide_empty)
+        self.extr = None
+
+    def build_mongo_filter(self, filter_expr):
+        referenced_tables = set()
+        for key, item in self.query.alias_refcount.items():
+            if item < 1:
+                continue
+            else:
+                referenced_tables.add(self.query.alias_map[key].table_name)
+
+        if len(referenced_tables) > 1:
+            raise NotImplementedError("Multi-table joins are not implemented yet.")
+
+        return MongoWhereNode(filter_expr, self.mongo_meta)
+
+    def get_distinct_clause(self):
+        return [
+            {
+                "$group": {
+                    "_id": {
+                        col.target.attname: f"${col.target.column}" for col, _, alias in self.select
+                    },
+                },
+            },
+            {"$replaceRoot": {"newRoot": "$_id"}},
+        ]
+
+    def as_operation(self, with_limits=True, with_col_aliases=False):
```

Comment:
MongoDB implementation of [`as_sql`](https://github.com/django/django/blob/main/django/db/models/sql/compiler.py#L1740) of the base sql compiler